### PR TITLE
Mso trigger

### PR DESCRIPTION
**Pull request by wsxrdv**

### CODE REVIEW

File: scaaml/plot.py [hunk 1]
```diff
@@ -86,6 +86,14 @@ def plot_comparison(traces, labels, title):
     plt.show()
 
 
+def plot_trace_and_trigger(trace, trigger, fig_filename="capture.pdf"):
```
Comment (jmichelp) [inline scaaml/plot.py:89]: PDF not PNG?

File: scaaml/capture/scope/ps6424e.py [hunk 2]
```diff
@@ -163,21 +172,45 @@ def channel(self):
...
+    @property
+    def port_pin(self) -> Optional[int]:
+        """Number of pin of the digital port."""
+        return self._port_pin
+
+    @port_pin.setter
+    def port_pin(self, val: Optional[int]) -> None:
+        # If the channel is digital port is not None.
+        if self.is_digital:
+            assert val is not None
```
Comment (jmichelp) [inline scaaml/capture/scope/ps6424e.py:195]: I would keep assertions for validating boundaries within our code. Here it takes user input so I think raising ValueError would give a more actionnable error to the user.

File: scaaml/capture/scope/ps6424e.py [hunk 3]
```diff
@@ -310,6 +351,15 @@ def _dict_repr(self):
         ret["trigger_direction"] = self.trigger_direction
         return ret
 
+    @property
+    def is_digital(self) -> bool:
+        """Return True iff this channel is a digital channel (PORT0 or PORT1).
```
Comment (jmichelp) [inline scaaml/capture/scope/ps6424e.py:356]: typo or is the double "f" voluntary?

File: scaaml/capture/scope/ps6424e.py [hunk 4]
```diff
@@ -532,7 +585,124 @@ def get_last_trace(self, as_int: bool = False) -> np.ndarray:
 
     def get_last_trigger_trace(self) -> np.ndarray:
         """Return a copy of the last trigger trace."""
-        return np.array(self._buffers[1][:])
+        # Digital trigger.
+        if self.trigger.is_digital:
+            bits = np.array(self._buffers[1][:], dtype=np.int16)
+            port_pin = self.trigger.port_pin
+            assert port_pin is not None
+            return np.array(bits & (2**port_pin), dtype=np.float32)
```
Comment (jmichelp) [inline scaaml/capture/scope/ps6424e.py:593]: Probably faster and more standard to replace the exponentiation by a bitshift here:
`2**port_pin` => `1 << port_pin`

File: scaaml/capture/scope/ps6424e.py [hunk 5]
```diff
@@ -532,7 +585,124 @@ def get_last_trace(self, as_int: bool = False) -> np.ndarray:
...
+        #    any transition on channel causes a trigger
+        directions = (picoStruct.DIGITAL_CHANNEL_DIRECTIONS *
+                      digital_channels)()
+        for i in range(digital_channels):
+            if i == self.trigger.port_pin:
+                directions[i] = picoStruct.DIGITAL_CHANNEL_DIRECTIONS(
+                    PICO_PORT_DIGITAL_CHANNEL[
+                        f"PICO_PORT_DIGITAL_CHANNEL{i}"],  # channel
+                    picoEnum.PICO_DIGITAL_DIRECTION[
+                        "PICO_DIGITAL_DIRECTION_RISING"],  # direction
```
Comment (jmichelp) [inline scaaml/capture/scope/ps6424e.py:689]: Maybe that's personal preference but I would factor the trigger direction here:
```python
for i in range(digital_channels):
  trig_direction = "PICO_DIGITAL_DONT_CARE"
  if i == self.trigger.port_pin:
    trig_direction = "PICO_DIGITAL_RISING"
  directions[i] = ....
```

File: scaaml/capture/scope/ps6424e.py (hunk 3 above)
Comment (kralka) [reply scaaml/capture/scope/ps6424e.py]: iff == if and only if (sorry my math talk :))

**Review by jmichelp — APPROVED** [review]

**Pull request merged**